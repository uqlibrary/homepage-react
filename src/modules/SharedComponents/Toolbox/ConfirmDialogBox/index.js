export { default as ConfirmationBox } from './components/ConfirmationBox';
