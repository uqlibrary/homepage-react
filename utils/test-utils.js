/* eslint-disable react/prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { mui1theme } from 'config/theme';
import { Provider } from 'react-redux';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';

import preview from 'jest-preview';

import { getStore } from '../src/config/store';
import Immutable from 'immutable';
import { createMemoryHistory } from 'history';

const domTestingLib = require('@testing-library/dom');
const reactTestingLib = require('@testing-library/react');

const { configure } = domTestingLib;

configure(config => ({
    ...config,
    testIdAttribute: 'id',
}));

const AllTheProviders = props => {
    return (
        <MuiThemeProvider theme={mui1theme}>
            <MuiPickersUtilsProvider utils={MomentUtils}>{props.children}</MuiPickersUtilsProvider>
        </MuiThemeProvider>
    );
};

AllTheProviders.propTypes = {
    children: PropTypes.node,
};

export const rtlRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

export const renderWithRouter = (
    ui,
    { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
    return render(
        <AllTheProviders>
            <Router history={history}>{ui}</Router>
        </AllTheProviders>,
    );
};

export const RenderWithRouter = ({
    children,
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
}) => (
    <AllTheProviders>
        <Router history={history}>{children}</Router>
    </AllTheProviders>
);
export const renderWithRedux = ({ initialState }) => render => {
    return {
        ...render,
        store: getStore({ initialState, history: render.history }),
    };
};

export const withRouter = ({
    route = '/',
    path = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
} = {}) => WrappedComponent => {
    return (
        <Router history={history}>
            <Route path={path} children={WrappedComponent} />
        </Router>
    );
};

export const withRedux = (initialState = Immutable.Map()) => WrappedComponent => {
    return <Provider store={getStore(initialState)}>{WrappedComponent}</Provider>;
};

// eslint-disable-next-line react/prop-types
export const WithRedux = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>{children}</Provider>
);

export const WithReduxStore = ({ initialState = Immutable.Map(), children }) => (
    <Provider store={getStore(initialState)}>
        <AllTheProviders>{children}</AllTheProviders>
    </Provider>
);

module.exports = {
    ...domTestingLib,
    ...reactTestingLib,
    rtlRender,
    renderWithRouter,
    renderWithRedux,
    withRedux,
    withRouter,
    WithRedux,
    WithReduxStore,
    RenderWithRouter,
    preview,
};
