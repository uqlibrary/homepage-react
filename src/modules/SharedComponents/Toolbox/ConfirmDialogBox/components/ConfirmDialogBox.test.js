import React from 'react';
import { ConfirmDialogBox, styles } from './ConfirmDialogBox';
import { rtlRender, fireEvent } from 'test-utils';

const getProps = (testProps = {}) => ({
    classes: {},
    hideCancelButton: false,
    locale: {
        alternateActionButtonLabel: 'Maybe',
        confirmationTitle: 'Confirmation',
        confirmationMessage: 'Are you sure?',
        cancelButtonLabel: 'No',
        confirmButtonLabel: 'Yes',
    },
    onAction: jest.fn(),
    onCancelAction: jest.fn(),
    onAlternateAction: jest.fn(),
    onRef: jest.fn(),
    showAlternateActionButton: false,
    isOpen: true,
    ...testProps,
});

function setup(testProps = {}) {
    return rtlRender(<ConfirmDialogBox {...getProps(testProps)} />);
}

describe('ConfirmDialogBox snapshots tests', () => {
    it('renders component with yes/no buttons', () => {
        const { getByText, getByRole } = setup();
        expect(getByText('Confirmation')).toBeInTheDocument();
        expect(getByText('Are you sure?')).toBeInTheDocument();
        expect(getByRole('button', { name: /Yes/ })).toBeInTheDocument();
        expect(getByRole('button', { name: /No/ })).toBeInTheDocument();
    });

    it('renders component with yes', () => {
        const { getByRole, queryByRole } = setup({ hideCancelButton: true });
        expect(getByRole('button', { name: /Yes/ })).toBeInTheDocument();
        expect(queryByRole('button', { name: /No/ })).not.toBeInTheDocument();
    });

    it('renders component with yes/no/maybe buttons', () => {
        const { getByRole } = setup({ showAlternateActionButton: true });
        expect(getByRole('button', { name: /Yes/ })).toBeInTheDocument();
        expect(getByRole('button', { name: /Maybe/ })).toBeInTheDocument();
        expect(getByRole('button', { name: /No/ })).toBeInTheDocument();
    });

    it('should have a proper style generator', () => {
        const theme = {
            breakpoints: {
                up: jest.fn(() => 'test1'),
            },
            palette: {
                warning: {
                    main: 'test2',
                    dark: 'test3',
                },
                white: {
                    main: 'test4',
                },
            },
        };
        expect(styles(theme)).toMatchSnapshot();

        delete theme.palette.warning;
        delete theme.palette.white;
        expect(styles(theme)).toMatchSnapshot();

        delete theme.palette;
        expect(styles(theme)).toMatchSnapshot();
    });

    it('renders component with customised locale', () => {
        const { getByRole, getByText } = setup({
            locale: {
                alternateActionButtonLabel: 'ENG: Maybe',
                cancelButtonLabel: 'ENG: No',
                confirmationTitle: 'ENG: Confirmation',
                confirmationMessage: 'ENG: Are you sure?',
                confirmButtonLabel: 'ENG: Yes',
            },
            showAlternateActionButton: true,
        });
        expect(getByText('ENG: Confirmation')).toBeInTheDocument();
        expect(getByText('ENG: Are you sure?')).toBeInTheDocument();
        expect(getByRole('button', { name: /ENG: Yes/ })).toBeInTheDocument();
        expect(getByRole('button', { name: /ENG: Maybe/ })).toBeInTheDocument();
        expect(getByRole('button', { name: /ENG: No/ })).toBeInTheDocument();
    });

    it('the ok-equivalent button should work', () => {
        const testFn = jest.fn();
        const { getByRole } = setup({
            onAction: testFn,
        });
        fireEvent.click(getByRole('button', { name: /Yes/ }));
        expect(testFn).toHaveBeenCalled();
    });

    it('the cancel-equivalent button should work', () => {
        const testFn = jest.fn();
        const { getByRole } = setup({
            onCancelAction: testFn,
        });
        fireEvent.click(getByRole('button', { name: /No/ }));
        expect(testFn).toHaveBeenCalled();
    });

    it('the alternate action button should work', () => {
        const testFn = jest.fn();
        const { getByRole } = setup({
            showAlternateActionButton: true,
            onAlternateAction: testFn,
        });
        fireEvent.click(getByRole('button', { name: /Maybe/ }));
        expect(testFn).toHaveBeenCalled();
    });

    describe('Class instance', () => {
        const getInstance = props => {
            const instance = new ConfirmDialogBox(props ?? getProps());
            instance.setState = jest.fn(newState => {
                instance.state = { ...instance.state, ...newState };
            });
            return instance;
        };

        it('should show and hide confirmation dialog', () => {
            const instance = getInstance({ ...getProps({ isOpen: false }) });
            instance.showConfirmation();
            expect(instance.state.isDialogOpen).toBeTruthy();

            instance._hideConfirmation();
            expect(instance.state.isDialogOpen).toBeFalsy();
        });
    });
});
