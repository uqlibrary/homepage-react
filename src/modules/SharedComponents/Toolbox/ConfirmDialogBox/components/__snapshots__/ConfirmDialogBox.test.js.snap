// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialogBox snapshots tests should have a proper style generator 1`] = `
Object {
  "alternateActionButtonClass": Object {
    "&:hover": Object {
      "backgroundColor": "test3",
    },
    "backgroundColor": "test2",
    "color": "test4",
  },
}
`;

exports[`ConfirmDialogBox snapshots tests should have a proper style generator 2`] = `
Object {
  "alternateActionButtonClass": Object {
    "&:hover": Object {
      "backgroundColor": undefined,
    },
    "backgroundColor": undefined,
    "color": undefined,
  },
}
`;

exports[`ConfirmDialogBox snapshots tests should have a proper style generator 3`] = `
Object {
  "alternateActionButtonClass": Object {
    "&:hover": Object {
      "backgroundColor": undefined,
    },
    "backgroundColor": undefined,
    "color": undefined,
  },
}
`;
